extensions/ql-vscode/src/debugger/debugger-ui.ts

```diff
@@ -0,0 +1,245 @@
+    this.ui.onSessionClosed(this.session);
+    this.dispose();
+  }
+
+  public async quickEval(): Promise<void> {
+    const args: CodeQLProtocol.QuickEvalRequest["arguments"] = {
+      quickEvalContext: await getQuickEvalContext(undefined),
+    };
+    await this.session.customRequest("codeql-quickeval", args);
+  }
+
+  /**
+   * Queues a message handler to be executed once all other pending message handlers have completed.
+   *
+   * The `onDidSendMessage()` function is synchronous, so it needs to return before any async
+   * handling of the msssage is completed. We can't just launch the message handler directly from
+   * `onDidSendMessage()`, though, because if the message handler's implementation blocks awaiting
+   * a promise, then another event might be received by `onDidSendMessage()` while the first message
+   * handler is still incomplete.
+   *
+   * To enforce sequential execution of event handlers, we queue each new handler as a `finally()`
+   * handler for the most recently queued message.
+   */
+  private queueMessageHandler(handler: () => Promise<void>): void {
+    this.lastDeferredMessageHandler =
+      this.lastDeferredMessageHandler.finally(handler);
+  }
+
+  /** Updates the UI to track the currently executing query. */
+  private async onEvaluationStarted(
+    body: CodeQLProtocol.EvaluationStartedEvent["body"],
+  ): Promise<void> {
+    const dbUri = Uri.file(this.configuration.database);
+    const dbItem = await this.dbm.createOrOpenDatabaseItem(dbUri);
+
+    // When cancellation is requested from the query history view, we just stop the debug session.
+    const tokenSource = new CancellationTokenSource();
+    tokenSource.token.onCancellationRequested(() =>
+      debug.stopDebugging(this.session),
+    );
+
+    this.localQueryRun = await this.localQueries.createLocalQueryRun(
+      {
+        queryPath: this.configuration.query,
+        quickEval: body.quickEvalContext,
+      },
+      dbItem,
+      new QueryOutputDir(body.outputDir),
+      tokenSource,
+    );
+  }
+
+  /** Update the UI after a query has finished evaluating. */
+  private async onEvaluationCompleted(
+    body: CodeQLProtocol.EvaluationCompletedEvent["body"],
+  ): Promise<void> {
+    if (this.localQueryRun !== undefined) {
+      const results: CoreQueryResults = body;
+      await this.localQueryRun.complete(results);
+      this.localQueryRun = undefined;
+    }
+  }
+}
+
+/** Service handling the UI for CodeQL debugging. */
+export class DebuggerUI
+  extends DisposableObject
+  implements DebugAdapterTrackerFactory
+{
+  private readonly sessions = new Map<string, QLDebugAdapterTracker>();
+
+  constructor(
+    private readonly app: App,
+    private readonly localQueryResultsView: ResultsView,
+    private readonly localQueries: LocalQueries,
+    private readonly dbm: DatabaseManager,
+  ) {
+    super();
+
+    if (isCanary()) {
+      this.push(debug.registerDebugAdapterTrackerFactory("codeql", this));
+    }
```

Comment:
This means that the user needs to restart VS Code when enabling the canary flag. Should we be handling changes to the canary flag here or is this a short-lived feature flag?

Reply:
It turns out that I can just remove the canary check here. There's no way to conditionally register a new debugger type, so the "codeql" debugger type is always registered (by `package.json`). Instead, there's code to test the canary flag and show an error if the user manages to create a "codeql" debug configuration and try to launch it. Since the debug session will fail to start, there's no harm in registering the debug adapter tracker unconditionally.